src/node/node_class.hpp

```diff
@@ -37,256 +57,1269 @@ using IndexPropertyType = js::IndexPropertyType<Types>;
+		}
+
+		status = napi_unwrap(env, instance, reinterpret_cast<void**>(&unwrapped));
+	}
+
+	NAPI_THROW_IF_FAILED(env, status, nullptr);
+	return unwrapped;
+}
 
-    return scope.Escape(instance);
+template<typename ClassType>
+inline typename ClassType::Internal* WrappedObject<ClassType>::get_internal() {
+	return m_internal.get();
 }
 
 template<typename ClassType>
-inline v8::Local<v8::FunctionTemplate> ObjectWrap<ClassType>::create_template() {
-    Nan::EscapableHandleScope scope;
+inline void WrappedObject<ClassType>::set_internal(Internal* internal) {
+	m_internal = std::unique_ptr<Internal>(internal);
+}
 
-    v8::Local<v8::FunctionTemplate> tpl = Nan::New<v8::FunctionTemplate>(construct);
-    v8::Local<v8::ObjectTemplate> instance_tpl = tpl->InstanceTemplate();
-    v8::Local<v8::String> name = Nan::New(s_class.name).ToLocalChecked();
+template<typename ClassType>
+void WrappedObject<ClassType>::set_factory_constructor(const Napi::Function& factoryConstructor) {
+	factory_constructor = Napi::Persistent(factoryConstructor);
+	factory_constructor.SuppressDestruct();
+}
 
-    tpl->SetClassName(name);
-    instance_tpl->SetInternalFieldCount(1);
+template<typename ClassType>
+Napi::Function WrappedObject<ClassType>::get_constructor(Napi::Env env) {
+	if (!constructor.IsEmpty()) {
+		return constructor.Value();
+	}
 
-    v8::Local<v8::FunctionTemplate> super_tpl = ObjectWrap<ParentClassType>::get_template();
-    if (!super_tpl.IsEmpty()) {
-        tpl->Inherit(super_tpl);
-    }
+	return Napi::Function(env, env.Null());
+}
 
-    // Static properties are setup in create_constructor() because V8.
-    for (auto &pair : s_class.static_methods) {
-        setup_static_method(tpl, pair.first, pair.second);
-    }
-    for (auto &pair : s_class.methods) {
-        setup_method(tpl, pair.first, pair.second);
-    }
-    for (auto &pair : s_class.properties) {
-        setup_property<v8::ObjectTemplate>(instance_tpl, pair.first, pair.second);
-    }
+template<typename ClassType>
+inline bool WrappedObject<ClassType>::is_instance(Napi::Env env, const Napi::Object& object) {
+	if (constructor.IsEmpty()) {
+		std::string typeName(typeid(ClassType).name());
+		std::string errorMessage = "is_instance: Class " + typeName + " not initialized. Call init() first";
+		throw Napi::Error::New(env, errorMessage);
+	}
+
+	Napi::HandleScope scope(env);
+
+	//Check the object is instance of the constructor. This will be true when the object have it's prototype set with setPrototypeOf. 
+	//This is true for objects configured in the schema with a function type.
+	Napi::Function ctor  = constructor.Value();
+	bool isInstanceOf = object.InstanceOf(ctor);
+	if (isInstanceOf) {
+		return true;
+	}
+
+	//Object store needs is_instance to return true when called with RealmObject instance even if the prototype was changed with setPrototypeOf
+	Napi::Object instance = object.Get("_instance").As<Napi::Object>();
```

Comment:
Would it make sense to store this using a JS Symbol instead? This way we don't get surprises if users of our API start using / override some of our "hidden" fields.

Reply:
Yes. will do a refactor

Reply:
resolving this. see comment below